CesiumGltf/include/CesiumGltf/MetadataConversions.h

```diff
@@ -210,64 +235,107 @@ struct MetadataConversions<
-   * This assumes that the entire std::string_view represents the number, not
+   * @brief Converts the contents of a std::string to an signed integer.
+   * This assumes that the entire std::string represents the number, not
    * just a part of it.
    *
    * This returns std::nullopt if no number is parsed from the string.
    *
-   * @param from The std::string_view to parse from.
+   * @param from The std::string to parse from.
    * @param defaultValue The default value to be returned if conversion fails.
    */
-  static std::optional<TTo> convert(const std::string_view& from) {
+  static std::optional<TTo> convert(const std::string& from) {
     if (from.size() == 0) {
       // Return early. Otherwise, empty strings will be parsed as 0, which is
       // misleading.
       return std::nullopt;
     }
 
-    // Amazingly, C++ has no* string parsing functions that work with strings
-    // that might not be null-terminated. So we have to copy to a std::string
-    // (which _is_ guaranteed to be null terminated) before parsing.
-    // * except std::from_chars, but compiler/library support for the
-    //   floating-point version of that method is spotty at best.
-    std::string temp(from);
+    if (from.find('-') != std::string::npos) {
+      // The string must be manually checked for a negative sign because for
+      // std::strtoull accepts negative numbers and bitcasts them, which is not
+      // desired!
+      return std::nullopt;
+    }
+
+    errno = 0;
 
     char* pLastUsed;
-    uint64_t parsedValue = std::strtoull(temp.c_str(), &pLastUsed, 10);
-    if (pLastUsed == temp.c_str() + temp.size()) {
-      // Successfully parsed the entire string as an integer of this type.
-      return CesiumUtility::losslessNarrow<TTo, uint64_t>(parsedValue);
+    uint64_t parsedValue = std::strtoull(from.c_str(), &pLastUsed, 10);
+    if (errno != EINVAL && errno != ERANGE) {
+      // Check if the entire string was parsed.
+      if (pLastUsed == from.c_str() + from.size()) {
+        // Successfully parsed the entire string as an integer of this type.
+        return CesiumUtility::losslessNarrow<TTo, uint64_t>(parsedValue);
+      }
     }
 
     // Failed to parse as an integer. Maybe we can parse as a double and
     // truncate it?
-    double parsedDouble = std::strtod(temp.c_str(), &pLastUsed);
-    if (pLastUsed == temp.c_str() + temp.size()) {
-      // Successfully parsed the entire string as a double.
-      // Convert it to an integer if we can.
-      double truncated = glm::trunc(parsedDouble);
-
-      uint64_t asInteger = static_cast<uint64_t>(truncated);
-      double roundTrip = static_cast<double>(asInteger);
-      if (roundTrip == truncated) {
-        return CesiumUtility::losslessNarrow<TTo, uint64_t>(asInteger);
+    errno = 0;
+
+    double parsedDouble = std::strtod(from.c_str(), &pLastUsed);
+    if (errno != EINVAL && errno != ERANGE) {
```

Comment:
From browsing the documentation on `strtod`, I do see ERANGE as one of the error values. I don't see EINVAL mentioned.

Also, I see that if no conversion can be performed, `0.0` or `HUGE_VAL` could be returned too.

Reply:
I'm still a little concerned that a conversion failure could happen here, but our code would interpret success.

From the [linux doc](https://linux.die.net/man/3/strtod),

```
Since 0 can legitimately be returned on both success and failure, the calling program should set errno to 0 before the call, and then determine if an error occurred by checking whether errno has a nonzero value after the call.
```

This hints that we should use something like...
```
if (errno == 0)
   // Check if the entire string was parsed.
   ...
```